nca/NetworkConfig/QueryOutputHandler.py

```diff
@@ -0,0 +1,397 @@
+            if self.conns_diff:
+                example_dict.update({'conns_config1': str(peers_conn.conns1), 'conns_config2': str(peers_conn.conns2)})
+            else:
+                example_dict.update({'conn': str(peers_conn.conns1)})
+            conns_lists.append(example_dict)
+
+        return [{'description': explanation_description, 'connections': conns_lists}]
+
+    def get_explanation_in_str(self, explanation_description):
+        """
+        returns the explanation result of ConnectionsDiffExplanation and its description in str.
+        When self.conns_diff is True, i.e. having conns1 and conns2 in PeersAndConnections items, the diff between
+        connection of each pair is printed
+        otherwise (having only conns1, connections from first config is printed)
+        :param str explanation_description: the relevant description of this output explanation
+        :rtype str
+        """
+        conns = []
+        for peers_conn in self.peers_diff_connections_list:
+            if self.conns_diff:
+                conns.append(f'src: {peers_conn.src_peer}, dst: {peers_conn.dst_peer}, description: '
+                             f'{peers_conn.conns1.print_diff(peers_conn.conns2, self.configs[0], self.configs[1])}')
+            else:
+                conns.append(f'src: {peers_conn.src_peer}, dst: {peers_conn.dst_peer}, conn: {peers_conn.conns1}')
+
+        return explanation_description + ':\n' + '\n'.join(conns) + '\n'
+
+
+@dataclass
+class OutputExplanation:
+    """
+    A class that unifies the possible types of the QueryAnswer's output_explanation as it may vary according to
+    the pattern of the query's result - an output_explanation may have only one of its fields at a time besides
+    to explanation_description
+    """
+    explanation_description: str = ''
+    policies_with_intersect_pods: IntersectPodsExplanation = None
+    policies_and_rules: PoliciesAndRulesExplanations = None
+    pods_lists: PodsListsExplanations = None
+    connections_diff: ConnectionsDiffExplanation = None
+    str_explanation: str = None  # for queries that compute separately the output in the required format
+    # (i.e. ConnectivityMapQuery, SanityQuery and SemanticDiffQuery)
+
+    def get_explanation_in_str(self):
+        """
+        computes and returns the output explanation is str, so it may be used to write the query answer in txt format.
+        An OutputExplanation object may have only one field representing the explanation data besides to
+         explanation_description, so the output is computed by calling the get_explanation_in_str method of this field
+         :return the explanation written with its description in txt format
+         :rtype: str
+        """
+        for val in self.__dict__.values():
+            if not isinstance(val, str) and val is not None:
+                return val.get_explanation_in_str(self.explanation_description)
+
```

Comment:
I think the function `get_explanation_in_str` is very dependent on the current fields of `OutputExplanation`, and this could be error-prone if someone in the future adds other fields to this class.
One option I could think of is having a base class `OutputExplanation` only with `explanation_description`, and letting the other classes (such as `IntersectPodsExplanation`) inherit from this class. You could also add a wrapper class for the `str_explanation`.
What do you think?

Reply:
yes, I agree , I'll try to re-organize the code with inheritance